### ISSUE TITLE
ability to serve temporary files from a session

### ISSUE THREAD

**Issue by gordonwoodhull** [att/rcloud#2436]
We've discussed this over and over, but I can't find an issue.

A session should be able to serve temporary files on an opt-in basis. It will register with a call like

```r
rcloud.serve.files('foo', '/tmp/bar')
```

And then the files will be accessible via something like `/proxy.R/{user}/{session}/bar`

It doesn't necessarily have to be `proxy.R` but it will use the same logic to find the session, so that seems like a logical place to put it. I'm not sure about the syntax: maybe there needs to be a magic "named port" to trigger the logic.

**Comment by gordonwoodhull** [comment]
I can see a really obvious way to implement this: `rcloud.serve_files()` registers another RCS key under `<user>/system/session/<id>` which contains the mapping of IDs to directories.

Then, if proxy.R detects a file request (eg port `:file`), it checks if it is the host. If so, it uses the mapping to deliver the file. If it's not the host, it forward its own proxy.R URL to the host (complete with port I guess).

The security risk it that's it's just RCS, so if someone obtains the session id of another running session, they can manipulate RCS and gain access to any publicly-readable files on that machine. That's not worse than what you can do with an ordinary notebook, so I think it's OK.

**Comment by s-u** [comment]
Well, if it's for a session this would be much easier if we just anchored it at the session directory - that one is well-defined so there would be no extra stuff to store, the lookup would be by session ID. In the current implementation the session directories are re-used, but we could very easily change them to be unique.

**Comment by s-u** [comment]
I guess the only issue would be that if this was to be enabled optionally, then there is still the security consideration of not using a flag in RCS. But hen it could simply be a magic file (`.expose` or so) in the session directory or something like that such that only the owner can enable it.

**Comment by gordonwoodhull** [comment]
Good ideas, much simpler. I guess we would just store the session directory in the existing session info.

Permissions are messy - I guess the session directory has to be readable by anyone (for `.expose` to work), and maybe `rcloud.serve.files` would create? or chmod? the named subdirectory making it readable by anyone.

Atm I'm looking for a solution that could be deployed in a package for RCloud 2.0, this may be too complicated though.

**Comment by gordonwoodhull** [comment]
No, we can't set the session directory in session info because that's RCS so it has the same security problems.

Are you imagining setting the session working directory based directly off the session id, like a hash or the string itself?

**Comment by s-u** [comment]
Yes, right now it's just using `Rconn`_xxx_, but since we are already setting up temp dirs and all, we could as well just setup the session directory. You are probably right that it may be better to use a hash - not that the session ID is all that hidden, but no need to make it extra easily discoverable. Note, however, that this approach (changing the session directory) does require a change in `rcloud.support` so can't be directly done in a package.

What could be done in a package is a way where the package will map the session ID to the directory itself, with the corresponding security implications.

(OT: on the long list we could think about access control in RCS - it's technically possible by using encryption. We could have session-level keys in addition to the obvious user level ones.)
